homeassistant/components/switch/sony_projector.py

```diff
@@ -0,0 +1,102 @@
+"""
+Use SDCP network control to power on/off and check state of Sony projector.
+
+For more details about this component, please refer to the documentation
+at https://home-assistant.io/components/switch.sony_projector/
+
+Example configuration:
+
+switch:
+  - platform: sony_projector
+    host: "192.168.1.43"
+
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.switch import (SwitchDevice, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    STATE_ON, STATE_OFF, CONF_NAME, CONF_HOST)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['pysdcp==1']
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'Sony Projector'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Connect with serial port and return Acer Projector."""
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+
+    add_entities([SonyProjector(host, name)], True)
+
+
+class SonyProjector(SwitchDevice):
+    """Represents an Acer Projector as a switch."""
+
+    def __init__(self, host, name):
+        """Init of the Sony projector."""
+        import pysdcp
+        self._sdcp = pysdcp.Projector(host)
```

Comment:
This should be part of `setup_platform()`.

Reply:
I'm not sure I agree with this one, but happy to be wrong. Doesn't the "SonyProjector" class abstract the underlying projector (Entity) from the platform (which may have multiple projectors, and the underlying pysdcp doesn't support that)

Reply:
Yes, the `SonyProjector` class abstracts the underlying projector from the platform. In the `setup_platform()` method, you should create a projector from `pydscp.Projector(host)` and pass that into the `SonyProjector` class' `__init__`.

Reply:
Sorry for being obtuse here - but what's the logic behind creating an instance of an internally required class outside of the Entity?

Incidentally, I followed the example in the "Acer Projector" switch, so I'd conform with the family code standard, and it defines the instance to the Serial object inside __init__?

Reply:
`setup_platform` is supposed to provide a success/failure (via returning `True`/`False`) to Home Assistant. There's no point in loading the Entity if we're unable to connect to the projector which we should know when we create the `Projector` instance, so Home Assistant can be informed that the platform is not ready.

That's my understanding at least, @fabaff may have some more reasons.

Reply:
Ah, OK - that makes a lot more sense. The underlying "pysdcp" library isn't that smart, but I can add additional code to do a first-pass test before returning from `setup_platform`.

Thanks for taking the time to explain :)